kliff/transforms/configuration_transforms/graphs/generate_graph.py

```diff
@@ -0,0 +1,199 @@
+        species,
+        cutoff,
+        n_layers,
+        as_torch_geometric_data=False,
+        implicit_fingerprint_copying=False,
+    ):
+        super().__init__(implicit_fingerprint_copying=implicit_fingerprint_copying)
+        self.species = species
+        self.cutoff = cutoff
+        self.n_layers = n_layers
+        self.infl_dist = n_layers * cutoff
+        self._tg = graph_module
+        if as_torch_geometric_data and not torch_geometric_available():
+            raise ImportError("Torch geometric is not available")
+        self.as_torch_geometric_data = as_torch_geometric_data
+
+    def forward(self, configuration: "Configuration"):
+        """
+        Generate a graph representation of a configuration.
+
+        Args:
+            configuration: Instance of ~:class:`kliff.dataset.Configuration`. For which the
+                graph representation is to be generated.
+
+        Returns:
+            C++ custom graph object or Pytorch Geometric Data object.
+        """
+        graph = graph_module.get_complete_graph(
+            self.n_layers,
+            self.cutoff,
+            configuration.species,
+            configuration.coords,
+            configuration.cell,
+            configuration.PBC,
+        )
+        graph.energy = configuration.energy
+        graph.forces = configuration.forces
+        if torch_available:
+            if self.as_torch_geometric_data:
+                return self._to_py_graph(graph)
+        return graph
+
+    @staticmethod
+    def _to_py_graph(graph):
+        """
+        Convert a C++ graph object to a KLIFF Geometric Graph Data object, ``GraphData``.
+
+        Args:
+            graph: C++ graph object.
+
+        Returns:
+            KLIFFTorchGraph object.
+        """
+        torch_geom_graph = KLIFFTorchGraph()
+        torch_geom_graph.energy = torch.as_tensor(graph.energy)
+        torch_geom_graph.forces = torch.as_tensor(graph.forces)
+        torch_geom_graph.n_layers = torch.as_tensor(graph.n_layers)
+        torch_geom_graph.coords = torch.as_tensor(graph.coords)
+        torch_geom_graph.images = torch.as_tensor(graph.images)
+        torch_geom_graph.species = torch.as_tensor(graph.species)
+        torch_geom_graph.z = torch.as_tensor(graph.z)
+        torch_geom_graph.contributions = torch.as_tensor(graph.contributions)
+        torch_geom_graph.num_nodes = torch.as_tensor(graph.n_nodes)
+        for i in range(graph.n_layers):
+            torch_geom_graph.__setattr__(
+                f"edge_index{i}", torch.as_tensor(graph.edge_index[i])
+            )
+        # torch_geom_graph.coords.requires_grad_(True)
+        return torch_geom_graph
+
+    def collate_fn_single_conf(self, config_list):
+        """
+        Collate function for use with a Pytorch DataLoader. This function is used when
+        dataloader loads each configuration as a batch of size 1. This is useful when
+        the model is trained on a single configuration at a time.
```

Comment:
Can we provide a general `collate_fn`? 

Although the current model might be trained using a single config, but in general we want to support any batch size.

Reply:
I am leaving this comment up for now. Because I need to investigate a bit more. A general `collate_fn` is part of base class, so that might not be needed, if I remember correctly this was needed for making the PyG collation properly. But I think we can remove it now. Perhaps in next commit.

Reply:
OK. Removed for now. If needed we can think of it later (when integrating unit tests).